stackstac/to_dask.py

```diff
@@ -196,36 +219,128 @@ def fetch_raster_window(
                     output = np.array(output)
                     all_empty = False
 
-                output[index] = data
+                band_i = asset_i_to_band[asset_i]
+                output[time_i, band_i : band_i + data.shape[0]] = data
 
     return output
 
 
 def normalize_chunks(
-    chunks: ChunksParam, shape: Tuple[int, int, int, int], dtype: np.dtype
-) -> Tuple[Tuple[int, ...], Tuple[int, ...], Tuple[int, ...], Tuple[int, ...]]:
+    chunks: ChunksParam,
+    shape: Tuple[int, int, int, int],
+    nbands_per_asset: tuple[int, ...],
+    dtype: np.dtype,
+) -> tuple[TBYXChunks, tuple[int, ...]]:
     """
     Normalize chunks to tuple of tuples, assuming 1D and 2D chunks only apply to spatial coordinates
 
     If only 1 or 2 chunks are given, assume they're for the ``y, x`` coordinates,
     and that the ``time, band`` coordinates should be chunksize 1.
+
+    If "auto" is given for bands, uses ``nbands_per_asset``.
+
+    Returns
+    -------
+    chunks:
+        Normalized chunks
+    asset_table_band_chunks:
+        Band chunks to apply to the asset table (see `process_multiband_chunks`)
     """
     # TODO implement our own auto-chunking that makes the time,band coordinates
     # >1 if the spatial chunking would create too many tasks?
     if isinstance(chunks, int):
-        chunks = (1, 1, chunks, chunks)
+        chunks = (1, nbands_per_asset, chunks, chunks)
     elif isinstance(chunks, tuple) and len(chunks) == 2:
-        chunks = (1, 1) + chunks
+        chunks = (1, nbands_per_asset) + chunks
+    elif isinstance(chunks, tuple) and len(chunks) == 4 and chunks[1] == "auto":
+        chunks = (chunks[0], nbands_per_asset, chunks[2], chunks[3])
 
-    return da.core.normalize_chunks(
+    norm: TBYXChunks = da.core.normalize_chunks(
         chunks,
         shape,
         dtype=dtype,
-        previous_chunks=((1,) * shape[0], (1,) * shape[1], (shape[2],), (shape[3],)),
+        previous_chunks=((1,) * shape[0], nbands_per_asset, (shape[2],), (shape[3],)),
         # ^ Give dask some hint of the physical layout of the data, so it prefers widening
         # the spatial chunks over bundling together items/assets. This isn't totally accurate.
     )
 
+    # Ensure we aren't trying to split apart multi-band assets. This would require rewriting
+    # the asset table (adding duplicate columns) and is generally not what you want, assuming
+    # that in multi-band assets, the bands are stored interleaved, so reading one requires reading
+    # them all anyway.
+    asset_table_band_chunks = process_multiband_chunks(norm[1], nbands_per_asset)
+    return norm, asset_table_band_chunks
+
+
+def process_multiband_chunks(
```

Comment:
There's a probably an easier/more legible way to implement this function, but this turned out to be the crux of implementing multi-band. We really don't want two different chunks to source data from the same multi-band asset (different bands to different chunks), because this would no longer be a blockwise operation (from asset table -> full array), therefore would require rewriting the asset table, opening the same dataset twice, and probably be bad performance anyway when bands are interleaved. It seemed easier to just validate that this situation doesn't occur than to support it.

If you have an asset with lots of bands which _aren't_ stored interleaved, I could definitely see wanting different chunks per band. I'm just not sure how common that is. It would be helpful to know if this seems like a common use-case @TomAugspurger.

Reply:
These docs describe the three methods for organizing multiple bands, two of which use interleaving: https://desktop.arcgis.com/en/arcmap/10.3/manage-data/raster-and-images/bil-bip-and-bsq-raster-files.htm

And a shorter synopsis of each: https://www.l3harrisgeospatial.com/docs/enviimagefiles.html
BIP is common for hyperspectral datasets like ASTER
BIL I think is the most common format. Landsat comes in BIL
BSQ seems to be less common, legacy sats like SPOT were distributed in BSQ: https://www.loc.gov/preservation/digital/formats/fdd/fdd000306.shtml but I'm not certain on how common BSQ is.